src/util/MemMapper/MemMapperUnix.cpp

```diff
@@ -42,6 +42,12 @@ namespace MemMapper
 
 	void* AllocateMemory(void* baseAddr, size_t size, PAGE_PERMISSION permissionFlags, bool fromReservation)
 	{
+		uint64 page_size = sysconf(_SC_PAGESIZE);
+		if ((uint64) baseAddr % page_size != 0)
+		{
+			baseAddr = (void*) ((uint64)baseAddr & ~(page_size - 1));
+		}
+
```

Comment:
Silently allocating a different range than the one requested is probably not the way to go. Wii U page size is actually 128KB, therefore 16KB pages shouldn't be an issue unless we accidentally hardcoded in a 4K requirement somewhere. Sorry for not properly communicating this before.

`mmuRange_HIGHMEM` is a bit special but the previous workaround was fine imho. But let me know if there were any other issues and we can probably find a way to make all systems compatible with 16KB page size.

Reply:
> Silently allocating a different range than the one requested is probably not the way to go. 

I didn't mean to put this for allocation, only for the `mprotect` call as the entire page has to have the same permissions, that's my mistake. I'll update it now.

> Wii U page size is actually 128KB, therefore 16KB pages shouldn't be an issue unless we accidentally hardcoded in a 4K requirement somewhere. 

Looking into WiiUBrew's documentation of the memory map, it seems that this memory space isn't allocated on it's own, as the entire area from `0xFFFF0000` to `0xFFFFFFFF`, which includes `mmuRange_HIGHMEM`, appears to be mapped at once.

> `mmuRange_HIGHMEM` is a bit special but the previous workaround was fine imho

I removed the previous workaround in favor of this fix because this should be more flexible for other page sizes other than 4KB and 16KB and wouldn't be Apple-specific

Reply:
> Looking into WiiUBrew's documentation of the memory map, it seems that this memory space isn't allocated on it's own, as the entire area from 0xFFFF0000 to 0xFFFFFFFF, which includes mmuRange_HIGHMEM, appears to be mapped at once.

You looked at the wrong memory map. The relevant PPC userspace one is on [this page](https://wiiubrew.org/wiki/Cafe_OS). But even there it's defined as starting from 0xFFFE0000 which is exactly one PPC page (128KB). This is a bit of a tangent, but the issue is that we don't really support emulating per-thread memory (yet) and the only reason we have defined it in Cemu at all is to prevent Project Zero from crashing since it writes to the very last byte of the address space due to a bug in the game's code. I don't want to define the memory range with the proper full size because that is still incorrect as it's not per-thread. Hence we limited it to as small a size possible so any unsupported access to it would be caught and not go through silently and lead to hard-to-debug crashes later on.

Regardless, I still believe that code using `ReserveMemory`, `AllocateMemory` and `FreeMemory` should be aware of page boundaries and that internally it shouldn't mess with the parameters or allow unaligned base or size. I know the existing code doesn't assert on this right now but ideally it would.